src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -837,26 +991,153 @@ public new CSharpCompilation ReplaceSyntaxTree(SyntaxTree oldTree, SyntaxTree ne
+                {
+                    var tree = _syntaxTrees[i];
+                    treesBuilder.Add(tree);
+                    ordinalMapBuilder.Add(tree, i);
+                }
+                AppendAllSyntaxTrees(treesBuilder, newTree, ordinalMapBuilder, ref loadDirectiveMap, ref declMap, ref declTable, ref referenceDirectivesChanged);
+                for (var i = oldOrdinal + 1; i < _syntaxTrees.Length; i++)
+                {
+                    var tree = _syntaxTrees[i];
+                    treesBuilder.Add(tree);
+                    ordinalMapBuilder.Add(tree, treesBuilder.Count);
+                }
+
+                newTrees = treesBuilder.ToImmutableAndFree();
+                ordinalMap = ordinalMapBuilder.ToImmutableDictionary();
+                ordinalMapBuilder.Free();
+                Debug.Assert(newTrees.Length == ordinalMap.Count);
+            }
+            else
+            {
+                AddSyntaxTreeToDeclarationMapAndTable(newTree, _options, this.IsSubmission, ref declMap, ref declTable, ref referenceDirectivesChanged);
+
+                Debug.Assert(ordinalMap.ContainsKey(oldTree)); // Checked by RemoveSyntaxTreeFromDeclarationMapAndTable
+
+                newTrees = _syntaxTrees.SetItem(oldOrdinal, newTree);
+
+                // CONSIDER: should this be an operation on ImmutableDictionary?
+                ordinalMap = ordinalMap.Remove(oldTree);
+                ordinalMap = ordinalMap.SetItem(newTree, oldOrdinal);
+            }
+
+            return UpdateSyntaxTrees(newTrees, ordinalMap, loadDirectiveMap, declMap, declTable, referenceDirectivesChanged);
+        }
+
+        /// <summary>
+        /// Removes all tree and any trees from #load'ed files.
+        /// </summary>
+        private int RemoveAllSyntaxTreesFromDeclarationMapAndTable(
+            SyntaxTree oldTree,
+            SyntaxTree newTree,
+            ref ImmutableDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMap,
+            ref ImmutableDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> declMap,
+            ref DeclarationTable declTable,
+            ref bool referenceDirectivesChanged,
+            ref bool loadDirectivesChanged)
+        {
+            var numTreesRemoved = 1;
+
+            ImmutableArray<LoadDirective> loadDirectives;
+            if (loadDirectiveMap.TryGetValue(oldTree, out loadDirectives))
+            {
+                Debug.Assert(loadDirectives.Any());
+                numTreesRemoved += loadDirectives.Length;
+                loadDirectiveMap = loadDirectiveMap.Remove(oldTree);
+                loadDirectivesChanged = HaveLoadDirectivesChanged(loadDirectives, newTree.GetCompilationUnitRoot().GetLoadDirectives());
+                if (loadDirectivesChanged)
+                {
+                    numTreesRemoved += RemoveAllSyntaxTreesFromDeclarationMapAndTableHelper(loadDirectives, ref loadDirectiveMap, ref declMap, ref declTable, ref referenceDirectivesChanged);
```

Comment:
Is numTreesRemoved incremented by loadDirectives.Length twice, once above and again in Helper method?